Blank page on photoshop.adobe.com when tapping in url bar, then pressing enter

cdumez opened the pull request:
#### 8d5ea3dad3eb952a9af07e1846f1632d42475729
<pre>
Blank page on photoshop.adobe.com when tapping in url bar, then pressing enter
<a href="https://bugs.webkit.org/show_bug.cgi?id=286771">https://bugs.webkit.org/show_bug.cgi?id=286771</a>
<a href="https://rdar.apple.com/143916165">rdar://143916165</a>

Reviewed by Darin Adler.

photoshop.adobe.com serves the `Cross-Origin-Opener-Policy: same-origin` and
`Cross-Origin-Embedder-Policy: require-corp` HTTP headers on responses in
order because they use SharedArrayBuffer. These headers may cause us to
process-swap when receiving such a response and the origin of the source
and destination differ.

In the reproduction case, the user would first load the photoshop URL,
which would serve those headers. They would then tap the URL bar and hit
Enter. This would cause Safari to load the same URL in a *new* WKWebView
backed by the same WebProcess (thanks to the _relatedWebView SPI).
Because the same process is used, the second load would find the main
resource in the memory cache and try to revalidate it from the server
(The response has an ETag header). We would send the conditional/revalidation
request to the server and the server would repond with a bodyless 304
HTTP response, meant to allow us to reuse the cached resource. However,
the 304 response also contained the COOP &amp; COEP headers. When seeing those
headers, we would decide to process-swap because the source origin is null
(the new WebView has not loaded anything yet) and is thus different than
the destination origin. We would spin a new WebProcess, swap to it and send
it the 304 response. Because the new WebProcess does not have the resource
in its memory cache and it is receiving a bodyless 304 response, it just
displays an empty document.

To address the issue, we no longer attempt to revalidate resources when
the load is a navigation and the cached response has a Cross-Origin-Opener-Policy
header other than &quot;unsafe-none&quot;, since those are likely to cause a process
swap and we wouldn&apos;t know how to swap in case of 304 response.

I have verified manually on photoshop.adobe.com that this addresses the issue
and this is covered by a new API test.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::determineRevalidationPolicy const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, COOPAndCOEPOn304Response)):
* Tools/TestWebKitAPI/cocoa/HTTPServer.h:
(TestWebKitAPI::HTTPResponse::setShouldRespondWith304ToConditionalRequests):
* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::statusText):
(TestWebKitAPI::serialize304Response):
(TestWebKitAPI::HTTPServer::respondToRequests):

Canonical link: <a href="https://commits.webkit.org/289631@main">https://commits.webkit.org/289631@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f320a984a8cc601956604779ee0396233108f2ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87562 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7077 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92432 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38306 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89613 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15245 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25382 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90564 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5700 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79241 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47988 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5486 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33634 "Found 60 new test failures: fast/repaint/align-self-change.html fast/repaint/align-self-overflow-change.html fast/repaint/emoji-glyph-overflow-repaint-fail.html fast/repaint/justify-items-legacy-change.html fast/repaint/justify-self-change.html fast/repaint/justify-self-overflow-change.html http/tests/security/contentSecurityPolicy/block-all-mixed-content/insecure-script-in-iframe.html http/tests/security/cookie-module-propagate.html http/tests/security/cookie-module.html http/tests/websocket/tests/hybi/client-close-2.html ... (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37419 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75892 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34499 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94313 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14730 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76467 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14984 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75090 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75690 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18615 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20071 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18495 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7681 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14746 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/20047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14490 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17934 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
```diff
@@ -325,6 +328,19 @@ static ASCIILiteral statusText(unsigned statusCode)
+        responseBuilder.append(pair.key, ": "_s, pair.value, "\r\n"_s);
+    responseBuilder.append("\r\n"_s);
+
+    return responseBuilder.toString().utf8().span();
```
A bummer that we first build a `StringImpl`, then a `CStringBuffer`, then a `Vector` all with the same characters. Would be nicer if we had a clean idiom that didn’t involve 3 copies of the string. I guess the `appendUTF8ToVector` function just below has the same issue with extra unnecessary copying, and we could fix that one too at some point.

WTF has the building blocks needed to figure out the size and bytes from a UTF-8 string so they can go directly into creating a `Vector` or appending to it, it’s just a tiny bit more code than just using `String::utf8()` so it’s a small simple project, but maybe wasteful if we have no reason to need to optimize.

cdumez replied:
I didn't spend much time optimizing since this is a test http server only used by API tests, not production code. But I'll look into doing better.

cdumez replied:
Proposing:
```
static Vector<uint8_t> toUTF8Vector(const String& string)
{
    Vector<uint8_t> result;
    string.tryGetUTF8([&](std::span<const char8_t> utf8) {
        result.append(byteCast<uint8_t>(utf8));
        return true;
    });
    return result;
}
```

darinadler replied:
I agree that it’s not important to optimize this test code. I’m sorry I pushed you in this direction!

But that does look better, pretty good. I still think there must be one unneeded copy there, because it had to allocate a buffer to do the conversion and we’d love to instead just figure out the size and decode right into the vector storage itself.

darinadler commented on Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
```diff
@@ -325,6 +328,19 @@ static ASCIILiteral statusText(unsigned statusCode)
 
+static Vector<uint8_t> serialize304Response(const HashMap<String, String>& headerFields)
+{
+    int statusCode = 304;
```
```suggestion
    constexpr int statusCode = 304;
```

darinadler commented on Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
```diff
@@ -368,6 +384,14 @@ static void appendUTF8ToVector(Vector<uint8_t>& vector, const String& string)
 
         auto response = requestData->requestMap.get(path);
+        if (response.shouldRespondWith304ToConditionalRequests) {
+            if (contains(request.span(), "If-None-Match:"_span)) {
```
This seems unnecessarily loose. It will match even if there is a character before "If", like if there was something that said "Not-If-None-Match:", or if this string appears somewhere else inside the request, like in the path of a URL perhaps. Is there a way to do this check more correctly?

cdumez replied:
Yes, I admit I could have done much better here. This is testing code so I hesitated. But given that you mentioned it, I'll improve it.

darinadler replied:
Maybe that’s not critical. I need to remember it’s a lower bar in the test code than in WebKit itself.

cdumez committed:
Blank page on photoshop.adobe.com when tapping in url bar, then pressing enter
https://bugs.webkit.org/show_bug.cgi?id=286771
rdar://143916165

Reviewed by Darin Adler.

photoshop.adobe.com serves the `Cross-Origin-Opener-Policy: same-origin` and
`Cross-Origin-Embedder-Policy: require-corp` HTTP headers on responses in
order because they use SharedArrayBuffer. These headers may cause us to
process-swap when receiving such a response and the origin of the source
and destination differ.

In the reproduction case, the user would first load the photoshop URL,
which would serve those headers. They would then tap the URL bar and hit
Enter. This would cause Safari to load the same URL in a *new* WKWebView
backed by the same WebProcess (thanks to the _relatedWebView SPI).
Because the same process is used, the second load would find the main
resource in the memory cache and try to revalidate it from the server
(The response has an ETag header). We would send the conditional/revalidation
request to the server and the server would repond with a bodyless 304
HTTP response, meant to allow us to reuse the cached resource. However,
the 304 response also contained the COOP & COEP headers. When seeing those
headers, we would decide to process-swap because the source origin is null
(the new WebView has not loaded anything yet) and is thus different than
the destination origin. We would spin a new WebProcess, swap to it and send
it the 304 response. Because the new WebProcess does not have the resource
in its memory cache and it is receiving a bodyless 304 response, it just
displays an empty document.

To address the issue, we no longer attempt to revalidate resources when
the load is a navigation and the cached response has a Cross-Origin-Opener-Policy
header other than "unsafe-none", since those are likely to cause a process
swap and we wouldn't know how to swap in case of 304 response.

I have verified manually on photoshop.adobe.com that this addresses the issue
and this is covered by a new API test.

* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::determineRevalidationPolicy const):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessSwapOnNavigation.mm:
((ProcessSwap, COOPAndCOEPOn304Response)):
* Tools/TestWebKitAPI/cocoa/HTTPServer.h:
(TestWebKitAPI::HTTPResponse::setShouldRespondWith304ToConditionalRequests):
* Tools/TestWebKitAPI/cocoa/HTTPServer.mm:
(TestWebKitAPI::statusText):
(TestWebKitAPI::serialize304Response):
(TestWebKitAPI::HTTPServer::respondToRequests):

Canonical link: https://commits.webkit.org/289631@main
